src/theory/datatypes/theory_datatypes_utils.cpp

```diff
@@ -377,6 +378,196 @@ bool checkClash(Node n1, Node n2, std::vector<Node>& rew)
   return false;
 }
 
+struct SygusToBuiltinTermAttributeId
+{
+};
+typedef expr::Attribute<SygusToBuiltinTermAttributeId, Node>
+    SygusToBuiltinTermAttribute;
+
+Node sygusToBuiltin(Node n)
+{
+  Assert(n.isConst());
+  std::unordered_map<TNode, Node, TNodeHashFunction> visited;
+  std::unordered_map<TNode, Node, TNodeHashFunction>::iterator it;
+  std::vector<TNode> visit;
+  TNode cur;
+  unsigned index;
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    it = visited.find(cur);
+    if (it == visited.end())
+    {
+      if (cur.getKind() == APPLY_CONSTRUCTOR)
+      {
+        if (cur.hasAttribute(SygusToBuiltinTermAttribute()))
+        {
+          visited[cur] = cur.getAttribute(SygusToBuiltinTermAttribute());
+        }
+        else
+        {
+          visited[cur] = Node::null();
+          visit.push_back(cur);
+          for (const Node& cn : cur)
+          {
+            visit.push_back(cn);
+          }
+        }
+      }
+      else
+      {
+        // non-datatypes are themselves
+        visited[cur] = cur;
+      }
+    }
+    else if (it->second.isNull())
+    {
+      Node ret = cur;
+      Assert(cur.getKind() == APPLY_CONSTRUCTOR);
+      const Datatype& dt = cur.getType().getDatatype();
+      // non sygus-datatype terms are also themselves
```

Comment:
Wouldn't it be better to treat them in the `else` above then? Not sure if they should be going in the cache as well, but if you think they should then you could just duplicate the attribute setting below.

Reply:
My reason for doing this was avoiding several things, the main being that I don't want to access the datatype of the node in two places (hence twice for each constructor node).  So I did intentionally do it this way.

Btw, it is a super rare case that we ever pay a penalty for the non-sygus datatype case (this would be any-constants over user-defined datatypes).  Hence we should optimize for the case where there is no non-sygus datatypes.

Reply:
Agreed. You could add a comment to have the reason for that choice documented in the code.